### PR TITLE
fix(indexer): asymmetric-pool entry threshold + decimals-unknown freshness gate (Lever 4 PR 1.6)

### PR DESCRIPTION
**Pull request by chapati23**
## Summary
- Persist effective threshold (10000 fallback) into breach `entryRebalanceThreshold` for asymmetric pools on the zero-threshold side; disable zero-entry heal when entry was already captured under the fallback. Prevents breach re-scoring when reserves flip sides. (codex P2 #3214513401, deferred from PR #366.)
- Don't advance `oracleTimestamp` / `oracleOk` in the `tokenDecimalsKnown=false` early-return; gate dashboard health computation on `hasHealthData` so untrusted-decimals pools degrade visibly. (codex P2 #3214513402, deferred from PR #366.)

## Test plan
- [ ] `pnpm test` clean in `indexer-envio/`
- [ ] `pnpm test` clean in `ui-dashboard/`
- [ ] `pnpm typecheck` clean in both
- [ ] `trunk fmt` + `trunk check` clean
- [ ] Branch deploy: `/deploy-indexer --no-promote` after #366 merges and this rebases to main
- [ ] Spot-check: breach on asymmetric-zero-side pool keeps entry threshold across reserve flip
- [ ] Spot-check: pool with untrusted decimals shows "no-data" not "OK" on homepage

---

> [!NOTE]
> **Medium Risk**
> Touches indexer breach/health accounting and dashboard health rendering logic; mistakes could misclassify pool status or distort breach history/uptime, though changes are well-scoped and covered by new characterization tests.
> 
> **Overview**
> Fixes deviation-breach history for *asymmetric pools* by persisting the **effective** rising-edge threshold (`10000` fallback) via new `breachEntryThreshold`, disabling mid-breach “heal-from-zero” rewrites, and updating close-time critical-duration scoring to use an explicit legacy `10000` floor.
> 
> Hardens *decimals-unknown* behavior by holding the indexer freshness cursor (`oracleTimestamp`/`oracleOk`/related fields) when `tokenDecimalsKnown=false`, adding `hasHealthData=false → "N/A"` gating in both indexer and dashboard `computeHealthStatus`, and ensuring `computeEffectiveStatus` doesn’t collapse `N/A`+`OK` into `OK`.
> 
> Cleans up UI breach-history components by removing the `pool` prop from `BreachHistoryChart`/table rows and canonicalizing legacy `entryRebalanceThreshold` fallbacks to `10000` (never live pool thresholds). Adds/updates tests around these edge cases and extends `react-doctor` suppressions for touched long/stateful components.
> 
> <sup>Reviewed by [Cursor Bugbot](https://cursor.com/bugbot) for commit 9cc61c2e3992d9d79d3afa789bf4d5d8b8f01942. Bugbot is set up for automated code reviews on this repo. Configure [here](https://www.cursor.com/dashboard/bugbot).</sup>

### CODE REVIEW

**Comment by chapati23** [comment]
@claude review

**Comment by chapati23** [comment]
cursor review

File: ui-dashboard/src/lib/health.ts [hunk 1]
```diff
@@ -202,13 +209,25 @@ export function isOracleFresh(
...
  */
 export function computeHealthStatus(
   pool: PoolHealthState,
   chainId?: number,
   nowSeconds: number = Math.floor(Date.now() / 1000),
 ): HealthStatus {
   if (pool.source?.includes("virtual")) return "N/A";
+  // Indexer flagged the deviation accrual as untrusted — don't render
+  // synthesized health status. See docblock for rationale.
+  if (pool.hasHealthData === false) return "N/A";
```
[the remark this answers is not in the document]
Comment (chapati23) [reply ui-dashboard/src/lib/health.ts]: Fixed in 431f220 — added `hasHealthData=false → N/A` short-circuit to indexer's `computeHealthStatus` mirror in `pool.ts`, with docblock noting it's defense-in-depth (the upstream `tokenDecimalsKnown` early-return in sortedOracles handlers already prevents reaching it on untrusted samples, but the function itself is now in parity). `makePool` fixture default flipped to `hasHealthData: true` so existing parity tests don't regress; tests exercising the no-data gate override explicitly.

File: indexer-envio/src/deviationBreach.ts [hunk 2]
```diff
@@ -177,19 +192,27 @@ export async function recordBreachTransition(
...
+    //
     // Conservative bound: time-resolved magnitude isn't tracked, so a breach
     // that briefly hit 1.05+ then dropped will still credit post-grace seconds.
     const healedEntryThreshold =
       (prev?.currentOpenBreachEntryThreshold ?? 0) > 0
         ? prev!.currentOpenBreachEntryThreshold
         : open.entryRebalanceThreshold > 0
           ? open.entryRebalanceThreshold
-          : next.rebalanceThreshold;
+          : Number(effectiveThreshold(next));
```
[the remark this answers is not in the document]
Comment (chapati23) [reply indexer-envio/src/deviationBreach.ts]: Fixed in 431f220 — the close-time fallback chain in `recordBreachTransition` now canonicalizes legacy `entry=0` directly to `10000` (named constant `LEGACY_ENTRY_THRESHOLD_FLOOR`) instead of `Number(effectiveThreshold(next))`. `breach-history-chart.tsx` UI fallback chain gets the same canonicalization (and the unused `pool` prop is dropped — tests updated). `nextOpenBreachEntryThreshold` doesn't need changes: fresh captures already use `breachEntryThreshold(next)` which never returns 0 for a real breach; legacy in-memory 0 just gets the floor at close-time via the entity-row path.

File: ui-dashboard/src/lib/health.ts [hunk 3]
```diff
@@ -400,6 +419,9 @@ export function computeEffectiveStatus(
     rebalanceThresholdAbove?: number;
     rebalanceThresholdBelow?: number;
     rebalanceThresholdsKnown?: boolean;
+    // Untrusted-deviation flag — propagated through `computeHealthStatus`.
+    // See `PoolHealthState.hasHealthData` for the gating semantics.
+    hasHealthData?: boolean;
```
[the remark this answers is not in the document]
Comment (chapati23) [reply ui-dashboard/src/lib/health.ts:438]: Fixed in 431f220 — `computeEffectiveStatus` now half-short-circuits on `N/A`: returns "N/A" when health is N/A AND limits are OK / N/A (your common case — homepage / OG paths degrade visibly instead of faking OK). When limits are WARN / CRITICAL / WEEKEND, those still flow through `worstStatus` because limit risk dominates health uncertainty by design. The PR 1.6 test `propagates the hasHealthData=false short-circuit through worstStatus` was updated to reflect the half-short-circuit semantic.

**Comment by chapati23** [comment]
@claude — addressed all review notes in 431f220 (merge of main + fixes):

**Cursor findings (CHANGES_REQUESTED, all fixed):**

- **Dashboard health gate not mirrored in indexer parity** → Fixed: indexer's `computeHealthStatus` gains `hasHealthData=false → "N/A"` short-circuit. `makePool` test fixture defaults to `hasHealthData: true` so parity tests don't regress.
- **Legacy `entryRebalanceThreshold = 0` mis-scored after side flip** → Fixed: close-time fallback in `recordBreachTransition` canonicalizes legacy `entry=0` directly to the 10000 floor instead of `effectiveThreshold(next)` (which would consult the post-flip live side). `breach-history-chart.tsx` UI fallback gets the same fix (and drops the now-unused `pool` prop).
- **`computeEffectiveStatus` still shows OK for no-data pools** → Fixed: half-short-circuits on N/A — returns "N/A" when health is N/A AND limits are OK/N/A (defeats the prior `worstStatus(N/A, OK) = OK` collapse), passes through to `worstStatus` for elevated limit states (WARN/CRITICAL/WEEKEND).

**Your minor items (non-blocking, both addressed):**

- **Mirror asymmetric case untested** → Added test for `breachEntryThreshold({rebalanceThresholdAbove: 300, rebalanceThresholdBelow: 0, ...})` alongside the existing `above=0, below=300` case.
- **Misleading test title `"10000 floor"`** → Test comment block updated; the title is now accurate because the close-time fallback chain literally canonicalizes to 10000 (cursor #3214689033 fix above).

Plus: PR #366 merged 5 min before this review, so this push also resolves the merge conflicts (BACKLOG.md, sortedOracles.ts, pool.ts, Test.ts, react-doctor.config.json, health.ts) by combining PR 1.6's freshness-cursor preserve + PR 1.5 round-12's never-rebalance exception, and keeps both `breachEntryThreshold` (PR 1.6) and `persistableThreshold` (PR 1.5 round-12) helpers — they target different write surfaces. Indexer 626 / dashboard 1986 passing.

File: ui-dashboard/src/lib/homepage-og.ts [hunk 4]
```diff
@@ -176,6 +176,44 @@ async function fetchChainSlice(network: Network): Promise<ChainSlice | null> {
...
+            ...p,
+            rebalanceThresholdAbove: t.rebalanceThresholdAbove,
+            rebalanceThresholdBelow: t.rebalanceThresholdBelow,
+            rebalanceThresholdsKnown: t.rebalanceThresholdsKnown,
+            tokenDecimalsKnown: t.tokenDecimalsKnown,
+          };
+    });
+  } catch {
+    // schema-lag / transient miss — pools render with under-bound fallback.
+  }
```
[the remark this answers is not in the document]
Comment (chapati23) [reply ui-dashboard/src/lib/homepage-og.ts]: Fixed in 90815a0 — leftover from the conflict resolution. The parallel `Promise.allSettled` block above (lines 137-148) already does the same fetch + merge; deleted the second sequential block.

File: indexer-envio/src/pool.ts [hunk 5]
```diff
@@ -200,12 +230,20 @@ export const isAboveCriticalMagnitude = (
...
+ *    first place — but mirroring the gate keeps the function pure and
+ *    answers any caller (parity tests, ad-hoc replay tooling) that
+ *    bypasses the upstream guard.
  */
 export function computeHealthStatus(
   pool: Pool,
   nowSeconds: bigint,
 ): IndexerHealthStatus {
   if (isVirtualPool(pool)) return "N/A";
+  if (pool.hasHealthData === false) return "N/A";
```
[the remark this answers is not in the document]
Comment (chapati23) [reply indexer-envio/src/pool.ts]: Fixed in 90815a0 — applied at all 5 caller sites: sortedOracles.OracleReported + .MedianUpdated, state-sync.UpdateReserves + .Rebalanced, limits-and-fees.RebalanceThresholdUpdated. Pattern: `merged = {...pool, ...poolUpdate, ...breachPoolUpdate}; persistedPool = {...merged, healthStatus: computeHealthStatus(merged, blockTimestamp)};` — the earlier healthStatus computed against pre-recordHealthSample hasHealthData would otherwise persist a stale N/A on the first-valid-sample edge.

File: ui-dashboard/src/lib/homepage-og.ts [hunk 6]
```diff
@@ -176,6 +176,44 @@ async function fetchChainSlice(network: Network): Promise<ChainSlice | null> {
     });
   }
 
+  // Threshold-known triple — isolated query so a schema-lag doesn't
+  // break the main pools fetch above. Fail-open: on miss the split
+  // fields stay undefined and `isNeverRebalance` returns false (safe
+  // under-bound). Without this merge, a 0/0 never-rebalance pool would
+  // appear in WARN/CRITICAL attention lists on the homepage OG card.
+  try {
+    const thresholdsRes = await client.request<{
```
[the remark this answers is not in the document]
Comment (chapati23) [reply ui-dashboard/src/lib/homepage-og.ts]: Fixed in 90815a0 — same finding as cursor #3214745363 above; the duplicate sequential block was leftover from the merge conflict resolution. Removed in this push.

File: indexer-envio/src/pool.ts (hunk 5 above)
[the remark this answers is not in the document]
Comment (chapati23) [reply indexer-envio/src/pool.ts]: Duplicate of #3214748736 above (codex appears to have double-posted). Fixed in 90815a0 — see resolution there.

File: ui-dashboard/src/components/breach-history-chart.tsx [hunk 7]
```diff
@@ -67,14 +66,15 @@ export function BreachHistoryChart({ breaches, pool }: Props) {
...
+    // re-score history (cursor #3214689033), so the legacy fallback
+    // canonicalizes to 10000 directly — the same under-bound the predicate
+    // scored against at rising edge for any pool legitimately captured at 0.
     const entryThreshold =
-      (b.entryRebalanceThreshold ?? 0) > 0
-        ? b.entryRebalanceThreshold!
-        : (pool.rebalanceThreshold ?? 0) > 0
-          ? pool.rebalanceThreshold!
-          : 10000;
+      (b.entryRebalanceThreshold ?? 0) > 0 ? b.entryRebalanceThreshold! : 10000;
```
[the remark this answers is not in the document]
Comment (chapati23) [reply ui-dashboard/src/components/breach-history-chart.tsx:77]: Fixed in 90815a0 — `breach-row.tsx` fallback chain now canonicalizes legacy entry=0 directly to 10000, mirroring the chart fix. The intermediate `pool.rebalanceThreshold` step is dropped (would consult the post-flip live side). Test `falls back to pool.rebalanceThreshold...` updated to assert the 10000 floor instead. Bonus cleanup: `pool` prop dropped from `BreachRow` + `BreachTable` signatures (now unused).

File: indexer-envio/src/deviationBreach.ts [hunk 8]
```diff
@@ -177,19 +192,31 @@ export async function recordBreachTransition(
...
     // Conservative bound: time-resolved magnitude isn't tracked, so a breach
     // that briefly hit 1.05+ then dropped will still credit post-grace seconds.
+    const LEGACY_ENTRY_THRESHOLD_FLOOR = 10000;
     const healedEntryThreshold =
       (prev?.currentOpenBreachEntryThreshold ?? 0) > 0
         ? prev!.currentOpenBreachEntryThreshold
         : open.entryRebalanceThreshold > 0
           ? open.entryRebalanceThreshold
-          : next.rebalanceThreshold;
+          : LEGACY_ENTRY_THRESHOLD_FLOOR;
```
[the remark this answers is not in the document]
Comment (chapati23) [reply indexer-envio/src/deviationBreach.ts]: Fixed in 90815a0 — dropped `prev.currentOpenBreachEntryThreshold` from the close-time resolution chain entirely. For fresh PR-1.6 rows it agrees with `open.entryRebalanceThreshold` (same rising-edge code path); for legacy rows the OLD heal-from-zero branch may have written a post-flip opposite-side threshold there, which would re-score history. Now the entity-row capture is the only legacy detector. Updated comment block in `deviationBreach.ts` documents why.

File: ui-dashboard/src/lib/health.ts [hunk 9]
```diff
@@ -409,7 +441,9 @@ export function computeEffectiveStatus(
   chainId?: number,
 ): HealthStatus {
   const health = computeHealthStatus(pool, chainId);
-  return worstStatus(health, resolveLimitStatus(pool));
+  const limit = resolveLimitStatus(pool);
+  if (health === "N/A" && (limit === "OK" || limit === "N/A")) return "N/A";
+  return worstStatus(health, limit);
```
[the remark this answers is not in the document]
Comment (chapati23) [reply ui-dashboard/src/lib/health.ts:450]: Fixed in 90815a0 — `global-pools-table.tsx` now calls `computeEffectiveStatus(p, network.chainId)` directly instead of inlining `worstStatus(computeHealthStatus, resolveLimitStatus)`. The hasHealthData N/A half-short-circuit (returns N/A when limits are OK/N/A, falls through to worstStatus when limits are WARN/CRITICAL/WEEKEND) now applies on the homepage table too. Dropped the unused `worstStatus` import.

File: indexer-envio/src/handlers/sortedOracles.ts [hunk 10]
```diff
@@ -135,7 +140,12 @@ SortedOracles.OracleReported.handler(async ({ event, context }) => {
       // accrual keeps moving on every oracle event even if decimals
       // self-heal stays stuck (e.g. governance-paused pool's fee tokens).
       if (!updatedPool.tokenDecimalsKnown && !isNeverRebalance(updatedPool)) {
-        context.Pool.set(updatedPool);
+        context.Pool.set({
+          ...updatedPool,
+          // Preserve freshness cursor — see comment above.
+          oracleTimestamp: existing.oracleTimestamp,
+          oracleOk: existing.oracleOk,
+        });
```
[the remark this answers is not in the document]
Comment (chapati23) [reply indexer-envio/src/handlers/sortedOracles.ts:155]: Fixed in 90815a0 — extended the OracleReported decimals-unknown freeze to preserve `oracleTxHash` AND `oraclePrice` alongside the existing timestamp/oracleOk preservation. Persisting the current event's tx hash next to a frozen timestamp would have rendered an inconsistent row in the oracle tab.

File: ui-dashboard/src/lib/health.ts [hunk 11]
```diff
@@ -202,13 +209,25 @@ export function isOracleFresh(
...
   chainId?: number,
   nowSeconds: number = Math.floor(Date.now() / 1000),
 ): HealthStatus {
   if (pool.source?.includes("virtual")) return "N/A";
+  // Indexer flagged the deviation accrual as untrusted — don't render
+  // synthesized health status. See docblock for rationale.
+  if (pool.hasHealthData === false) return "N/A";
   const isOracleStale = !isOracleFresh(pool, nowSeconds, chainId);
   if (isOracleStale) {
     // Distinguish expected weekend staleness from a real incident
```
[the remark this answers is not in the document]
Comment (chapati23) [reply ui-dashboard/src/lib/health.ts:234]: Fixed in 90815a0 — swapped guard order in UI `computeHealthStatus`: `isOracleStale → CRITICAL/WEEKEND` now runs BEFORE `hasHealthData=false → N/A`. Stale-oracle is an alertable freshness incident; must not be masked by the data-trust gate. Test `hasHealthData=false overrides what would otherwise be CRITICAL` flipped to assert CRITICAL wins (renamed accordingly).

File: indexer-envio/src/handlers/sortedOracles.ts [hunk 12]
```diff
@@ -386,11 +396,25 @@ SortedOracles.MedianUpdated.handler(async ({ event, context }) => {
...
+        context.Pool.set({
+          ...withThreshold,
+          // Preserve freshness cursor — see comment above.
+          oracleTimestamp: existing.oracleTimestamp,
+          oracleOk: existing.oracleOk,
+          // `lastOracleReportAt` is the median freshness anchor used by
+          // the indexer-side `derive` path; preserving it too keeps the
+          // anchor and the cursor in lockstep until self-heal lands.
+          lastOracleReportAt: existing.lastOracleReportAt,
+        });
```
[the remark this answers is not in the document]
Comment (chapati23) [reply indexer-envio/src/handlers/sortedOracles.ts:439]: Fixed in 90815a0 — extended the MedianUpdated decimals-unknown freeze to preserve `oracleTxHash` AND `oraclePrice` alongside the existing cursor (oracleTimestamp/oracleOk/lastOracleReportAt). Same rationale as #3214756049 for OracleReported.

File: indexer-envio/src/handlers/sortedOracles.ts (hunk 10 above)
[the remark this answers is not in the document]
Comment (chapati23) [reply indexer-envio/src/handlers/sortedOracles.ts:155]: Fixed in 90815a0 — see #3214756049 above (same fix covers both displayed-price + tx-hash preservation in OracleReported's freeze branch).

File: indexer-envio/src/pool.ts [hunk 13]
```diff
@@ -200,12 +230,20 @@ export const isAboveCriticalMagnitude = (
...
   if (!pool.oracleOk) return "CRITICAL";
```
[the remark this answers is not in the document]
Comment (chapati23) [reply indexer-envio/src/pool.ts:250]: Fixed in 90815a0 — swapped guard order in indexer `computeHealthStatus` (pool.ts mirror): `!oracleOk → CRITICAL` now runs BEFORE `hasHealthData=false → N/A`. Mirrors the UI fix from #3214756051.

**Comment by chapati23** [comment]
@claude — addressed 12 unreplied review comments in 90815a0 (rounds 1+2 of bot review on PR #370):

**Cursor (1 finding):**
- Duplicate trust-flag fetch in `homepage-og.ts` → Fixed: removed leftover sequential block from conflict resolution; parallel `Promise.allSettled` already covers it.

**Codex round 1 (6 findings — 1 duplicate-posted):**
- Recompute health after recordHealthSample (#3214748736 + #3214748739) → Fixed at 5 caller sites; persistedPool now reflects post-flip hasHealthData.
- Remove duplicate trust-flag fetch (#3214748738) → Fixed (same as cursor's).
- Breach chart vs breach-row fallback parity (#3214748742) → Fixed: breach-row canonicalizes legacy entry=0 to 10000 directly; bonus pool-prop dropdown cleanup.
- Ignore legacy healed denorms (#3214748744) → Fixed: dropped `prev.currentOpenBreachEntryThreshold` from the close-time chain; honor only the immutable entity-row capture.
- Apply N/A guard to pool table render (#3214748745) → Fixed: `global-pools-table` now uses `computeEffectiveStatus` directly.

**Codex round 2 (5 findings — all from the same review):**
- Preserve tx hash with frozen cursor (#3214756049 + #3214756054) → Fixed: OracleReported freeze now preserves oracleTxHash + oraclePrice.
- Preserve median tx hash (#3214756053) → Fixed: same for MedianUpdated.
- Oracle-staleness alerts under hasHealthData (#3214756051 + #3214756056) → Fixed: swapped guard order in UI + indexer mirror so `!oracleOk → CRITICAL` runs BEFORE `hasHealthData=false → N/A`.

Indexer 626 / dashboard 1986 passing; trunk clean.
